File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
from collections import deque

def solution(players, m, k):
    answer = 0
    servers = deque()
    n = len(players)
    for i in range(n):
        need = players[i] // m
                
        while servers:
            server = servers.popleft()
            if server > i:
                servers.appendleft(server)
                break
        
        current = len(servers)
        if current < need:
            answer += need - current
            for j in range(need-current):
                servers.append(i+k)

    return answer

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.ArrayList;
import java.util.PriorityQueue;

class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        int N = players.length;
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        for (int i = 0; i < N; i++) {
            int user = players[i];
            int cnt = pq.size();
            if (user >= (cnt+1) * m) {
                int need = user / m;
                
                answer += need - cnt;
                for (int j = 0; j < need-cnt; j++) {
                    pq.offer(k);
                }
            }
            
            ArrayList<Integer> tmp = new ArrayList<>();
            while (!pq.isEmpty()) {
                int cur = pq.poll();
                if (cur - 1 <= 0) {
                    continue;
                }
                tmp.add(cur-1);
            }
            
            for (int j = 0; j < tmp.size(); j++) {
                pq.offer(tmp.get(j));
            }
            
        }
        return answer;
    }
}